z3/src/goal.rs

```diff
@@ -0,0 +1,162 @@
+    pub fn assert(&self, ast: &impl ast::Ast<'ctx>) {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_assert(self.ctx.z3_ctx, self.z3_goal, ast.get_z3_ast())
+        }
+    }
+
+    /// Return true if the given goal contains the formula `false`.
+    pub fn is_inconsistent(&self) -> bool {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_inconsistent(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+
+    /// Return the depth of the given goal. It tracks how many transformations were applied to it.
+    pub fn get_depth(&self) -> u32 {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_depth(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+
+    /// Return the number of formulas in the given goal.
+    pub fn get_size(&self) -> u32 {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_size(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+
+    /// Return the number of formulas, subformulas and terms in the given goal.
+    pub fn get_num_expr(&self) -> u32 {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_num_exprs(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+
+    /// Return true if the goal is empty, and it is precise or the product of a under approximation.
+    pub fn is_decided_sat(&self) -> bool {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_is_decided_sat(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+    /// Return true if the goal contains false, and it is precise or the product of an over approximation.
+    pub fn is_decided_unsat(&self) -> bool {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_is_decided_unsat(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+
+    /// Erase all formulas from the given goal.
+    pub fn reset(&self) {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            Z3_goal_reset(self.ctx.z3_ctx, self.z3_goal)
+        };
+    }
+
+    /// Copy a goal `g` from the context `source` to the context `target`.
+    pub fn translate<'dest_ctx>(self, ctx: &'dest_ctx Context) -> Goal<'dest_ctx> {
+        Goal {
+            ctx,
+            z3_goal: unsafe {
+                let guard = Z3_MUTEX.lock().unwrap();
+                Z3_goal_translate(self.ctx.z3_ctx, self.z3_goal, ctx.z3_ctx)
+            }
+        }
+    }
+
+    /// Return the "precision" of the given goal. Goals can be transformed using over and under approximations.
+    pub fn get_precision(&self) -> GoalPrec {
+        let guard = Z3_MUTEX.lock().unwrap();
+        unsafe {
+            Z3_goal_precision(self.ctx.z3_ctx, self.z3_goal)
+        }
+    }
+
+    /// Return a vector of the formulas from the given goal.
+    pub fn get_formulas<T>(&self) -> Vec<T> where T: Ast<'ctx> {
```

Comment:
What about implementing this as an iterator?

Something like this:

```rust
fn iter_formulas(&self) -> impl Iterator<Item = T> {
  (0..goal_size).iter().map(|i| { lock(); Z3_goal_formula(...); })
}
```

It's up to you, just curious what your thoughts on this approach are

Reply:
That definitely seems like a better option though, since it can be lazily evaluated.  I'm running into some issues with the lifetime but will attempt to implement it.